gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy.publisher;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.configuration.WorkUnitState;
+import gobblin.data.management.copy.CopySource;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.data.management.util.PathUtils;
+import gobblin.publisher.DataPublisher;
+import gobblin.util.HadoopUtils;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+
+
+/**
+ * A {@link DataPublisher} to {@link CopyableFile}s from task output to final destination
+ */
+@Slf4j
+public class CopyDataPublisher extends DataPublisher {
+
+  private Path writerOutputDir;
+  private FileSystem fs;
+
+  public CopyDataPublisher(State state) throws IOException {
+    super(state);
+    Configuration conf = new Configuration();
+    String uri = this.state.getProp(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, ConfigurationKeys.LOCAL_FS_URI);
+
+    this.fs = FileSystem.get(URI.create(uri), conf);
+    this.writerOutputDir = new Path(state.getProp(ConfigurationKeys.WRITER_OUTPUT_DIR));
+  }
+
+  @Override
+  public void publishData(Collection<? extends WorkUnitState> states) throws IOException {
+
+    /*
+     * Multimap maps a dataset destination path to all workunits that belong to this dataset.
+     */
+    Multimap<Path, WorkUnitState> datasetRoots = getDatasetRoots(states);
+
+    for (Path datasetDestinationRoot : datasetRoots.keySet()) {
+      this.publishDataset(datasetDestinationRoot, datasetRoots.get(datasetDestinationRoot));
+    }
+
+  }
+
+  /**
+   * Create a multimap that maps a dataset destination path to all workunits that belong to this dataset.
+   */
+  private Multimap<Path, WorkUnitState> getDatasetRoots(Collection<? extends WorkUnitState> states) throws IOException {
```

Comment:
Why aren't the datasets explicitly serialized in the state? Then we don't have to have this logic to try to "infer" the roots. Further, the current API looks unusual. Why does the method return a multimap, when all that matters are the keys. Why not return a Dataset set?